python/src/spark_rapids_ml/core.py

```diff
@@ -1041,7 +1061,69 @@ def _transform_udf(pdf_iter: Iterator[pd.DataFrame]) -> pd.DataFrame:
                         data[pred.model_index] = index
                     yield data
 
-        return dataset.mapInPandas(_transform_udf, schema=schema)  # type: ignore
+        df = dataset.mapInPandas(_transform_udf, schema=schema)  # type: ignore
+        return self._try_stage_level_scheduling(df)
+
+    def _try_stage_level_scheduling(self, df: DataFrame) -> DataFrame:
+        """Try to enable stage-level scheduling for transform tasks"""
+        if self._skip_stage_level_scheduling():
+            return df
+
+        ss = _get_spark_session()
+
+        # Default the concurrent transform tasks to 2
+        concurrent_transform_tasks = int(
+            ss.conf.get("spark.rapids.ml.concurrentPredictTasks", "2")  # type: ignore
+        )
+
+        if concurrent_transform_tasks <= 0:
+            raise ValueError(
+                "The value of spark.rapids.ml.concurrentPredictTasks must be greater than 0."
+            )
+
+        executor_cores = int(ss.sparkContext.getConf().get("spark.executor.cores"))  # type: ignore
+
+        if (
+            executor_cores // 2 < concurrent_transform_tasks < executor_cores
+            or concurrent_transform_tasks > executor_cores
+        ):
+            original_concurrent_transform_tasks = concurrent_transform_tasks
+            concurrent_transform_tasks = (
+                executor_cores // 2
+                if concurrent_transform_tasks < executor_cores
+                else executor_cores
+            )
+            self.logger.warning(
+                "The spark.rapids.ml.concurrentPredictTasks should be in "
+                f"[1, {executor_cores // 2}] or {executor_cores}, "
+                f"we found it's {original_concurrent_transform_tasks}. But we're "
+                f"going to use {concurrent_transform_tasks}"
+            )
+
+        task_cores = executor_cores // concurrent_transform_tasks
+
+        default_task_cpus = int(ss.sparkContext.getConf().get("spark.task.cpus"))  # type: ignore
+        if task_cores == default_task_cpus:
+            self.logger.info(
+                "The Spark configurations have been set up to ensure that "
+                f"{concurrent_transform_tasks} transform tasks will be concurrently executed."
+            )
+            return df
+
+        treqs = TaskResourceRequests().cpus(task_cores).resource("gpu", 1.0)
+        rp = ResourceProfileBuilder().require(treqs).build
+
+        self.logger.info(
+            f"Transform tasks require the resource(cores={task_cores}, gpu=1.0) and there will be "
+            f"{concurrent_transform_tasks} transform tasks running at the same time."
+        )
+
+        # TODO Do we need to coalesce partitions to num_workers*concurrent_transform_tasks
+        # num_workers = self._infer_num_workers()
+        # df = df.coalesce(num_workers * concurrent_transform_tasks)
+
+        df = df.rdd.withResources(rp).toDF()
```

Comment:
Is there a perf impact of df -> rdd -> df  re. Spark RAPIDS plugin?

Reply:
Maybe should log a warning (w/ recommended settings) if there is an expected perf impact?

Reply:
Cool, Added the warning in the log.